### PR TITLE
Implement `From<bool>` for f32, f64

### PR DESCRIPTION
**Pull request by jhpratt**
As is required for trait implementations, these are insta-stable. Given there is a release tomorrow and this needs FCP, I set 1.65 as the stable version.

@rustbot label +A-floating-point +C-feature-request +needs-fcp +relnotes +S-waiting-on-review +T-libs-api -T-libs

### CODE REVIEW

**Comment by scottmcm** [comment]
Cross-linking the Zulip conversation: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/.60impl.20From.3Cbool.3E.20for.20f.7B32.7C64.7D.60/near/292687425

> Given there is a release tomorrow and this needs FCP, I set 1.65 as the stable version.

Reminder you can check https://github.com/rust-lang/rust/blob/master/src/version -- it's already been 1.65 for 5 days.

**Comment by thomcc** [comment]
I'm not sure the behavior here is intuitive enough to justify these. I'd have expected `0.0` to be the only falsey bool (this matches how many other languages do it, although sometimes they also include NaN), wheras in this implementation, values like 0.5 will be considered falsey (anything that truncates to zero).

I kind of feel like I prefer if people need to be explicit about this conversion.

**Comment by Lokathor** [comment]
You might be thinking of these backward? The input is a `bool`, you *couldn't* pass `0.5` to these functions.

In this impl: `true` is 1.0, and `false` is 0.0, and there's no other possible values that could be passed.

**Comment by jhpratt** [comment]
Yeah, this goes _from_ a boolean _to_ a float. The other way around would absolutely be unwise to implement in this manner.

**Comment by thomcc** [comment]
Oh, right. Yes, my bad. I don't have an issue then (this is what I get for taking a look at a PR at 11:30PM...)

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by jhpratt** [comment]
Rebased.

**Comment by clarfonthey** [comment]
I'm a bit torn on this. I think that if we do merge this, then directly casting from `bool` to floats should be allowed as part of the language. It feels inconsistent otherwise.

Note that this isn't against this feature necessarily, I'm just not sure what would justify having one of `as` or `From` work here without the other if the conversion is lossless.

**Comment by jhpratt** [comment]
While I understand what you're saying, I think the general trend is to move away from `as` as much as possible. So I think it would be best if we _didn't_ implement that cast, as it would be counter to other efforts.

**Comment by clarfonthey** [comment]
I would be in favour of that if there were actually detailed plans on how we plan to do that, but at least for now, this will be merged long before any of those plans are executed.

I mean ultimately it probably doesn't matter that much, but I feel like it's worth at least providing a decision on the `as` implementation in this decision, whether it's implemented or not.

Is the actual conversion more complicated than other `as` implementations? Was there a good reason why it wasn't implemented before, or did just no one care? If someone made a PR implementing this as an `as` conversion, would it be merged?

Again, it's very simple to merge this change, but it's also very simple to do the conversion through `u8` yourself. So what's the motivation there?

**Comment by Lokathor** [comment]
There is not a single, unified plan for moving away from `as`. It is still a general design guideline that many people, both on various teams and also those not on any teams, have tried to follow when possible. I myself agree that even if we can't replace `as` entirely right now, we should at least not expand it amy farther.

The main thing keeping `as` alive in my mind is that it's available in const, while `from` is not.

**Comment by clarfonthey** [comment]
Basically the key point I'm trying to make is that right now, since `as` exists and isn't going away any time soon, we should take agreeing that this `From` implementation is okay as *also* an agreement that the equivalent `as` implementation is also okay.

It doesn't mean that anyone is obliged to write the code for such an implementation before this is merged, but rather that if a compelling argument exists to *not* accept an `as` implementation for this other than "I don't like as", it should also be used as a reason against this change. Basically: why doesn't an `as` implementation exist right now, and is that reason actually a good design decision that can be applied here too?

**Comment by jhpratt** [comment]
> Was there a good reason why it wasn't implemented before, or did just no one care?

No idea. I haven't looked into the part of the compiler to see if there were/are any technical reasons it wasn't done (I suspect there aren't).

> I myself agree that even if we can't replace as entirely right now, we should at least not expand it any farther.

This is basically my view.

> we should take agreeing that this `From` implementation is okay as also an agreement that the equivalent as implementation is also okay.

I disagree. Just because there's not a concrete plan to move away from `as` doesn't mean we should continue expanding its capabilities. 

> is that reason actually a good design decision that can be applied here too

`as` in general is problematic. This is in part because the origin and target types can change without the author even realizing it. I'm not going to go into a ton of detail, as I trust that you're aware of the drawbacks of `as`.

**Comment by clarfonthey** [comment]
Oh, I absolutely would prefer to move away from `as` (at least as it exists right now) in favour of `From`. Not wanting to expand its capabilities is a perfectly valid reason for someone not going out of their way to implement it, but I don't think it's valid to reject someone else's implementation if they provide one.

My main thought process is that if it was specifically done because the operation itself is more expensive than the other `as` conversions, that might be a valid reason to require someone to write `f32::from(u8::from(val))` instead of just `f32::from(val)` where `val: bool`, similar to how the `as` implementation has to do `val as u8 as f32`.

But if the real answer is "well, we just didn't consider it," then I'm all down to continue with this change.

**Comment by jhpratt** [comment]
> I don't think it's valid to reject someone else's implementation if they provide one.

Agree to disagree on this point :shrug: I would personally opposite it, but I understand where you're coming from.

> if … the operation itself is more expensive

Considering bool → u8 is just reinterpreting the bytes, I couldn't imagine this (bool → f32) is more expensive than u8 → f32.

> But if the real answer is "well, we just didn't consider it,"

I suspect this is the case.

**Comment by dtolnay** [comment]
I am having some trouble envisioning the intended use of this impl in real code. Would anyone be able to point out code where using this impl would be the most desirable way to express what is meant?

@rfcbot concern needs use case

**Comment by jhpratt** [comment]
cc @repi — I know it was an internal feature request

**Comment by repi** [comment]
in graphics and shaders (using [rust-gpu](https://github.com/embarkstudios/rust-gpu)) this is pretty common to convert from a boolean expression to floats to enabling/disabling and scaling things binary.

and it is more explicit and less error prone than the previous `as` casting we were using and removed, and compiles both to native code and GPU code.

our current workaround, which is fine just not as nice as to have it built in:

```rust
/// Converts `bool` straight to `f32`.
///
/// Ideally Rust should support `f32::from(bool)`, but in the meantime this is a nicer short hand than
/// having to do `f32::from(i16::from(bool))`, or a more error prone lossy cast with `as i32 as f32` that we used to do
///
/// Rust support issue: <https://github.com/rust-lang/rust/issues/74015>
pub fn f32_bool(b: bool) -> f32 {
    // in Rust CPU code we would do this, but for rust-gpu that would use 16-bit
    // integer operations and fail with:
    // error: u16 without OpCapability Int16
    //
    // f32::from(i16::from(b))
    //
    // so do "lossy" cast that is 32-bit
    #[allow(clippy::cast_lossless)]
    {
        b as u32 as f32
    }
}
```

couple of random examples straight from our code base:

```rust
    let vel_diff = f32_bool(
        ((vel_now - vel_prev.xy()) / Vec2::ONE.max(abs_vec2(vel_now + vel_prev.xy()))).length()
            > 0.2,
    );
```

```rust
    quad_validity.x *= f32_bool(
        all_above_equal(bilinear_at_prev.px0(), 0)
            && all_below(bilinear_at_prev.px0(), out_tex_size),
    );
    quad_validity.y *= f32_bool(
        all_above_equal(bilinear_at_prev.px1(), 0)
            && all_below(bilinear_at_prev.px1(), out_tex_size),
    );
```

**Comment by dtolnay** [comment]
Thanks! @rfcbot resolve needs use case

**Comment by dtolnay** [comment]
Thanks!
@rfcbot resolve needs use case

File: library/core/src/convert/num.rs
Comment (riking) [inline library/core/src/convert/num.rs:172]: Remember to switch since to CURRENT_RUSTC_VERSION.

**Comment by jhpratt** [comment]
Rebased and updated stabilization to "current".

@rustbot ready

**Review by dtolnay — APPROVED** [review]

**Comment by jhpratt** [comment]
@dtolnay This needs to complete FCP, no?

**Comment by jhpratt** [comment]
Closing and reopening to avoid bors merging this for the moment, as I believe FCP is still necessary. This will also cancel FCP, but that can be restarted by any libs-api team member (checking the same boxes as are currently checked).

**Comment by dtolnay** [comment]
I do not feel that blocking this on FCP is necessary. The FCP proposal has been active for 4 months which is ample time for any team member to register their concern. In general the automation around FCP does not attempt to take into account size of the feature. The amount of process it makes sense to have around adding a trait impl to an existing type, or adding const to an existing method, is vastly different compared to adding some major language feature like async or const generics. The team has discretion to adapt the process to the situation.

**Pull request merged**
